target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata

```diff
@@ -192,6 +192,13 @@ case "$FIRMWARE" in
 	;;
 "ath10k/pre-cal-pci-0000:00:00.0.bin")
 	case $board in
+	comfast,cf-e313ac)
+		ath10kcal_extract "art" 20480 12064
+		ath10kcal_patch_mac_crc $(mtd_get_mac_binary art 4098)
```

Comment:
The decimal values in the two lines above should be updated to hex values (this has been changed treewide recently, just have a look at the other devices in the file now).

Reply:
True. I hadn't seen this changes, it makes sense. Fixed and rebased master.

Reply:
Please also change the second line: 4098 -> 0x1002

Reply:
Sorry, I should have checked more carefully myself.

Reply:
BTW: Have you checked that there actually is a (correct) MAC address in this location? I'm just asking because this location normally is used by &wmac, which you do not have here. So it looks odd that the vendor would put the address there.
Just want to make sure ...

Reply:
Well, there is a valid MAC, indeed. In &art (mtd1), I have:
```
0x0000: 40 A5 EF 2E 45 40
0x0006: 40 A5 EF 2E 45 42
*0x1002: 40 A5 EF 2E 45 41*
0x5006: 40 A5 EF 2E 45 43
```
but it's true that it's there right after a bunch of non-zero, non-0xFF bytes that I can't really decipher, and it's not as common a position as 0x0. Actually, to be consistent with the stock firmware, I should be using the one ending in :42, so I'll be replacing it. I think I copied this from another COMFAST device where the MACs were sequentially assigned as eth0-wlan0-eth1 instead of eth0-eth1-wlan0 like here.

Anyway, the thing is that reverted to the stock firmware and I found something weird. This is the correspondence in the stock firmware (a custom OpenWrt, by the way):
```
40 A5 EF 2E 45 40 => eth0 => in br-wan => physical Ethernet WAN port
40 A5 EF 2E 45 41 => eth1 => in br-lan => physical Ethernet LAN port
40 A5 EF 2E 45 42 => wlan0
40 A5 EF 2E 45 43 => N/A
```

But then I realised that the MAC addresses in the physical ports didn't match. That was because the stock firmware's `/etc/config/network` file does this:
```
config interface 'lan'
	option ifname 'eth1'
	option force_link '1'
	option type 'bridge'
	option proto 'static'
	option ipaddr '192.168.10.1'
	option netmask '255.255.255.0'
	option ip6assign '60'
	option macaddr '40:a5:ef:2e:45:40'

config eth1 'def_lan'

config interface 'wan'
	option ifname 'eth0'
	option proto 'dhcp'
	option type 'bridge'
	option macaddr '40:a5:ef:2e:45:41'

config eth0 'def_wan'
```
resulting in:
```
root@COMFAST:~# ifconfig  | grep "40:A5:EF:2E:45"
br-lan    Link encap:Ethernet  HWaddr 40:A5:EF:2E:45:40  
br-wan    Link encap:Ethernet  HWaddr 40:A5:EF:2E:45:41  
eth0      Link encap:Ethernet  HWaddr 40:A5:EF:2E:45:40  
eth1      Link encap:Ethernet  HWaddr 40:A5:EF:2E:45:41  
wlan0     Link encap:Ethernet  HWaddr 40:A5:EF:2E:45:42

┌─────────────────────────────────────────────────────────────┐
│                            Stock                            │
├──────────────────────────────┬──────────────────────────────┤
│ LAN port                     │ WAN port                     │
│ (40:a5:ef:2e:45:40)          │ (40:a5:ef:2e:45:41)          │
├───⇕──────────────────────────┼───⇕──────────────────────────┤
│ br-lan                       │ br-wan                       │
│ 40:a5:ef:2e:45:40 (forced)   │ 40:a5:ef:2e:45:41 (forced)   │
├───⇑──────────────────────────┼───⇑──────────────────────────┤
│ eth1                         │ eth0                         │
│ 40:a5:ef:2e:45:41 (assigned) │ 40:a5:ef:2e:45:40 (assigned) │
└──────────────────────────────┴──────────────────────────────┘
```
This is, eth0 is in br-wan, but br-wan is forced to use eth1's MAC, and the other way round.

I understand that, when supporting a device with OpenWrt, MAC addresses, device names and interface roles should be kept coherent with the stock firmware. However, what should I do here? Do I have to keep this weird MAC assignment to the bridges? Ideally, I would use one of this approaches:
```
┌───────────────────────────────────────────────────────────────┐
│                          OpenWrt (1)                          │
├───────────────────────────────┬───────────────────────────────┤
│ LAN port                      │ WAN port                      │
│ (40:a5:ef:2e:45:40)           │ (40:a5:ef:2e:45:41)           │ => same
├───⇕───────────────────────────┼───⇕───────────────────────────┤
│ br-lan                        │ br-wan                        │
│ 40:a5:ef:2e:45:40 (inherited) │ 40:a5:ef:2e:45:41 (inherited) │ => same
├───⇑───────────────────────────┼───⇑───────────────────────────┤
│ eth0                          │ eth1                          │ => swapped
│ 40:a5:ef:2e:45:40 (assigned)  │ 40:a5:ef:2e:45:41 (assigned)  │ => same
└───────────────────────────────┴───────────────────────────────┘
┌───────────────────────────────────────────────────────────────┐
│                          OpenWrt (2)                          │
├───────────────────────────────┬───────────────────────────────┤
│ LAN port                      │ WAN port                      │
│ (40:a5:ef:2e:45:40)           │ (40:a5:ef:2e:45:41)           │ => same
├───⇕───────────────────────────┼───⇕───────────────────────────┤
│ br-lan                        │ br-wan                        │
│ 40:a5:ef:2e:45:40 (inherited) │ 40:a5:ef:2e:45:41 (inherited) │ => same
├───⇑───────────────────────────┼───⇑───────────────────────────┤
│ eth1                          │ eth0                          │ => same
│ 40:a5:ef:2e:45:40 (assigned)  │ 40:a5:ef:2e:45:41 (assigned)  │ => swapped
└───────────────────────────────┴───────────────────────────────┘
```
but I am not sure what's the best way to go. Any suggestions will be very welcome :)

Reply:
Thank you very much for your detailed analysis.
No matter what I'd like to say about people like the developers of this vendor, I would do the following:
1. The ports should keep the same state (i.e. wan/lan) as in stock firmware.
2. We cannot easily change what's eth0 and eth1 on the running, so the assignment to wan and lan is determined by 1.
3. I would slightly prefer to base MAC assignment on what the ports are doing. This means the address found on the stock WAN port should be the address set for the OpenWrt WAN port. The port modes will be what most of the user care about, while the internal assignment of eth0/eth1 is hidden to the typical user. However, note the word "slightly" here, we could also have thrown a coin to decide.

So, assuming that the current `ucidef_set_interfaces_lan_wan "eth0" "eth1"` sets the port modes correctly, I think this would be:
eth0 0x0
eth1 0x1002
wlan 0x6

Reply:
And BTW: Have you checked what happens if you do not set the MAC address in 11_fix_wifi_dings?
Normally, if you read caldata from 0x5000, it might automatically read the the MAC address from 0x5006 (which is present). So, as for the ethernet, it might be the case the radio is initialized with a valid address which is then overwritten by another one ...
(You do not need to check that, unless you are curious...)

Reply:
Hmm, it's even more complicated:
`ucidef_set_interfaces_lan_wan "eth0" "eth1"` is referring to ethX on the running device, so what I wrote refers to the running device, too:
eth0 0x0 (=lan)
eth1 0x1002 (=wan)
wlan 0x6

In DTS, eth0/eth1 are swapped, so we have
&eth0 0x1002 (=wan)
&eth1 0x0 (=lan)

So, based on your current commit state, you need to change &eth0 to 0x1002 and the WiFi setup to 0x6 ...

(This still assumes that your ports are correctly assigned to WAN and LAN...)

Reply:
> Thank you very much for your detailed analysis.

Thank **you** for reviewing the PR! :hugs: 

> No matter what I'd like to say about people like the developers at this vendor, I would do the following:
> 
>     1. The ports (!) should keep the same state (i.e. wan/lan) as in stock firmware.
:+1: That's for granted, specially since they're labeled as LAN and WAN.

> 
>     2. We cannot easily change what's eth0 and eth1 on the running device, so the assignment to wan and lan is determined by 1.
Well, kind of. I can use either the current configuration in the DTS file:
```
&eth0 {
	status = "okay";
	phy-handle = <&swphy0>;
	mtd-mac-address = <&art 0xXXX>;

	gmac-config {
		device = <&gmac>;
		switch-phy-swap = <1>;
	};
};

&eth1 {
	status = "okay";
	mtd-mac-address = <&art 0xYYY>;
};
```
or this one which also works and does not swap eth0 and eth1:
```
&eth0 {
	status = "okay";
	phy-handle = <&swphy4>;
	mtd-mac-address = <&art 0xZZZ>;
};

&eth1 {
	status = "okay";
	mtd-mac-address = <&art 0x0>;
	mtd-mac-address = <&art 0xTTT>;
};
```
(then, of course, setting XXX/YYY/ZZZ/TTT to the correct values, and also 01_leds and 02_network).

Both work; which one is preferred? I'd rather stick with the first one because I like the "default" eth0=>LAN eth1=>WAN, but I have no other reasoning than that.

>     3. I would slightly prefer to base MAC assignment on what the ports are doing. This means the address found on the stock WAN port should be the address set for the OpenWrt WAN port. The port modes will be what most of the user care about, while the internal assignment of eth0/eth1 is hidden to the typical user. However, note the word "slightly" here, we could also have thrown a coin to decide.
:+1:  Same MACs on the physical ports as the stock firmware, I agree, is a must. Beyond that... I flipped a coin already in 2. :smiley: 

> So, assuming that the current ucidef_set_interfaces_lan_wan "eth0" "eth1" sets the port modes > correctly, I think this would be:
> eth0 0x0
> eth1 0x1002
> wlan 0x6
>
>Hmm, it's even more complicated:
> `ucidef_set_interfaces_lan_wan "eth0" "eth1"` is referring to ethX on the running device, so what I wrote refers to the running device, too:
> eth0 0x0 (=lan)
> eth1 0x1002 (=wan)
> wlan 0x6
> 
> In DTS, eth0/eth1 are swapped, so we have
> &eth0 0x1002 (=wan)
> &eth1 0x0 (=lan)
> 
> So, based on your current commit state, you need to change &eth0 to 0x1002 and the WiFi setup to 0x6 ...
> 
> (This still assumes that your ports are correctly assigned to WAN and LAN...)

That's right. Anyway, whatever ends up regarding point 2, I'll make sure the stock Ethernet port - MAC assignment is maintained.




> And BTW: Have you checked what happens if you do not set the MAC address in 11_ath10k_caldata?
> Normally, if you read caldata from 0x5000, it might automatically read the the MAC address from 0x5006 (which is present). So, as for the ethernet, it might be the case the radio is initialized with a valid address which is then overwritten by another one ...
> (You do not need to check that, unless you are curious...)

It gets the mac at 0x5006 (40:A5:EF:2E:45:43) which, :open_mouth: surprise!, it's not the one used by the stock firmware :smile:

Reply:
If you swap eth0 and eth1 with switch-phy-swap, are you sure that both ports work in both cases? From my experience, only one of the settings made it possible to have both ports working, while with the other one only one port worked. So, in the end there was only one possible case.

Reply:
> If you swap eth0 and eth1 with switch-phy-swap, are you sure that both ports work in both cases? From my experience, only one of the settings made it possible to have both ports working, while with the other one only one port worked. So, in the end there was only one possible case.

Hmmm... I'll double check and report.

Reply:
> If you swap eth0 and eth1 with switch-phy-swap, are you sure that both ports work in both cases? From my experience, only one of the settings made it possible to have both ports working, while with the other one only one port worked. So, in the end there was only one possible case.

If I'm not doing anything wrong, https://github.com/rogerpueyo/openwrt/tree/ath79-cf-e313ac and https://github.com/rogerpueyo/openwrt/tree/ath79-cf-e313ac-no-swap are equivalent. In the first one I get eth0 for LAN, eth1 for WAN. In the second one I get eth0 for WAN, eth1 for LAN.

In both cases I'm able to use LAN/WAN correctly, iperf3 bandwidth is the same (~95 Mbps) so, unless if I'm missing something, I can't tell any practical difference.